torch/fx/experimental/symbolic_shapes.py

```diff
@@ -265,6 +266,29 @@ def check_consistent(new, old) -> None:
         assert isinstance(old, scalar_types) and not isinstance(old, bool), f"{old} != {new}"
         torch._check(old == new, lambda: f"{old} != {new} (old != new)")
 
+def rebind_unbacked(shape_env, n: torch.fx.Node, result):
+    """
+    Suppose we are retracing a pre-existing FX graph that previously had
+    fake tensor propagation (and therefore unbacked SymInts).  When we retrace,
+    we re-propagate fake tensors, which results in new unbacked SymInts.
+    When this happens, we need to tell the shape environment about the equivalence
+    of the old and new unbacked SymInts.  Pass us the old torch.fx.Node (which
+    has the old binding information) and the new result (which we can extract the
+    new unbacked SymInts out from).
+    """
+    if bindings := n.meta.get("unbacked_bindings"):
+        for raw_u0, path in bindings.items():
+            u1 = pytree.key_get(result, path)
+            # We should never have bindings for raw bools; instead they should
+            # have been converted to ints via ConvertIntKey
+            assert type(u1) is not bool
+            if isinstance(u1, (int, float)):
+                raw_u1 = sympy.sympify(u1)
+            else:
+                raw_u1 = u1.node.expr
+            # TODO: replace with rename unbacked to
+            shape_env.defer_runtime_assert(sympy.Eq(raw_u0, raw_u1), "")
```

Comment:
more of a curiosity question / me trying to understand, but: I imagined that `raw_u1` is the new expression that we want to statically "ignore", and declare equivalent to the old expression `raw_u0`. Why does that involve adding a runtime assert? (or does `defer_runtime_assert` exit early here)

Reply:
Yeah, this is the TODO. It ends up getting fixed in https://github.com/pytorch/pytorch/pull/124316